internal/topology/upgrade/clusterctl_upgrade_test.go

```diff
@@ -0,0 +1,1174 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get referenced actualCluster.spec.controlPlaneRef")
+	}
+	addToAllObj(cluster.Name, ".spec.controlPlaneRef", refObj)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}] = refObj.GetGeneration()
+	if additionalChecks != nil {
+		if err := additionalChecks(refObj); err != nil {
+			return nil, errors.Wrap(err, "failed additional checks on actualCluster.spec.controlPlaneRef")
+		}
+	}
+
+	cpInfraRef, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(refObj)
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster controplPlane.spec.machineTemplate.infrastructureRef is not yet set")
+	}
+	refObj, err = getReferencedObject(ctx, env.GetClient(), cpInfraRef, version, actualCluster.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get referenced controplPlane.spec.machineTemplate.infrastructureRef")
+	}
+	addToAllObj(cluster.Name, "controlPlane.spec.machineTemplate.infrastructureRef", refObj)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}] = refObj.GetGeneration()
+
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	if err = env.List(ctx, machineDeployments, client.InNamespace(cluster.Namespace), client.MatchingLabels{
+		clusterv1.ClusterNameLabel:          cluster.Name,
+		clusterv1.ClusterTopologyOwnedLabel: "",
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to list machineDeployments")
+	}
+	if len(machineDeployments.Items) != 1 {
+		return nil, errors.Errorf("expected 1 machineDeployment, got %d", len(machineDeployments.Items))
+	}
+
+	for _, md := range machineDeployments.Items {
+		refs[clusterv1.ContractVersionedObjectReference{
+			APIGroup: md.GroupVersionKind().Group,
+			Kind:     md.GroupVersionKind().Kind,
+			Name:     md.GetName(),
+		}] = md.GetGeneration()
+		addToAllObj(cluster.Name, "machineDeployment "+md.Name, &md)
+
+		refObj, err = getReferencedObject(ctx, env.GetClient(), &md.Spec.Template.Spec.InfrastructureRef, version, actualCluster.Namespace)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get referenced machineDeployment.spec.template.spec.infrastructureRef")
+		}
+		addToAllObj(cluster.Name, "machineDeployment "+md.Name+" .spec.template.spec.infrastructureRef", refObj)
+		refs[clusterv1.ContractVersionedObjectReference{
+			APIGroup: refObj.GroupVersionKind().Group,
+			Kind:     refObj.GetKind(),
+			Name:     refObj.GetName(),
+		}] = refObj.GetGeneration()
+		if additionalChecks != nil {
+			if err := additionalChecks(refObj); err != nil {
+				return nil, errors.Wrap(err, "failed additional checks on machineDeployment.spec.template.spec.infrastructureRef")
+			}
+		}
+
+		if md.Spec.Template.Spec.Bootstrap.ConfigRef == nil {
```

Comment:
I think this should never happen as we create the MD with the ref. If you want to keep it I guess we should do the same for MD infraRef above

Reply:
MD infraRef is not a pointer

Reply:
Ah I see, that explains the difference